core/templates/dev/head/pages/exploration_player/PlaythroughServiceSpec.js

```diff
@@ -0,0 +1,248 @@
+      this.ps.initSession(this.expId, this.expVersion);
+    }));
+
+    it('should initialize a session with correct values.', function() {
+      var playthrough = this.ps.getPlaythrough();
+
+      expect(playthrough.expId, this.expId);
+      expect(playthrough.expVersion, this.expVersion);
+    });
+
+    it('should record exploration start action.', function() {
+      this.ps.recordExplorationStartAction('initStateName1');
+      var playthrough = this.ps.getPlaythrough();
+
+      expect(playthrough.actions).toEqual(
+        [this.laof.createFromBackendDict({
+          action_type: 'ExplorationStart',
+          action_customization_args: {
+            state_name: {
+              value: 'initStateName1'
+            }
+          },
+          schema_version: 1
+        })]);
+    });
+
+    it('should record answer submit action.', function() {
+      this.ps.recordAnswerSubmitAction(
+        'stateName1', 'stateName2', 'TextInput', 'Hello', 'Try again', '30');
+      var playthrough = this.ps.getPlaythrough();
+
+      expect(playthrough.actions).toEqual(
+        [this.laof.createNew(
+          'AnswerSubmit', {
+            state_name: {
+              value: 'stateName1'
+            },
+            dest_state_name: {
+              value: 'stateName2'
+            },
+            interaction_id: {
+              value: 'TextInput'
+            },
+            answer: {
+              value: 'Hello'
+            },
+            feedback: {
+              value: 'Try again'
+            },
+            time_spent_in_state_secs: {
+              value: '30'
+            }
+          }, 1
+        )]);
+    });
+
+    it('should record exploration quit action.', function() {
+      this.ps.recordExplorationQuitAction('stateName1', '120');
+      var playthrough = this.ps.getPlaythrough();
+
+      expect(playthrough.actions).toEqual(
+        [this.laof.createNew(
+          'ExplorationQuit', {
+            state_name: {
+              value: 'stateName1'
+            },
+            time_spent_in_state_secs: {
+              value: '120'
+            }
+          }, 1
+        )]);
+    });
+
+    it(
+      'should analyze a playthrough for multiple incorrect submissions issue',
+      function() {
+        this.ps.recordExplorationStartAction('stateName1');
+        this.ps.recordAnswerSubmitAction(
+          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', '30');
+        this.ps.recordAnswerSubmitAction(
+          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', '30');
+        this.ps.recordAnswerSubmitAction(
+          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', '30');
+        this.ps.recordAnswerSubmitAction(
+          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', '30');
+        this.ps.recordAnswerSubmitAction(
+          'stateName1', 'stateName1', 'TextInput', 'Hello', 'Try again', '30');
```

Comment:
Do these calls need to be called right here? Seems like they belong more w/ the the second set of `expect` calls (just before line 122). Same elsewhere.

Reply:
Done.